src/target/riscv/riscv-013.c

```diff
@@ -686,6 +696,27 @@ static int dmi_write_exec(struct target *target, uint32_t address,
 	return dmi_op(target, NULL, NULL, DMI_OP_WRITE, address, value, true, ensure_success);
 }
 
+static bool check_dbgbase_exists(struct target *target)
+{
+	uint32_t dbgbase_to_find = target->dbgbase;
+	/* Start with the first debug module */
+	uint32_t nextdm = 0;
+
+	LOG_TARGET_DEBUG(target, "Searching for DM with DMI base address (dbgbase) = 0x%x", target->dbgbase);
+	while (1) {
+		target->dbgbase = nextdm;
```

Comment:
Can you rewrite this function not to change target at all? Using `target->dbgbase` as a temporary variable, only to restore it at the end, makes the code hard to follow.

Reply:
Change `target->dbgbase` is because `dmi_read`(which eventually call `dmi_scan`) will use ` target->dbgbase` to access the target DM

Reply:
I see. I think that highlights a problem with this implementation, where layers of abstraction are mixed together where they really ought to be separate.

Reply:
I need to go through all the DMs starting at 0.
The current implementation is to pass the `target` to `dmi_read`, so I need to keep changing the `target->dbgbase`.
If change to pass `dm` to `dmi_read`, then I need to keep changing `dm->dbgbase`.
I'll try that.

Reply:
done